atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/TimestampLockAwareTransaction.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.transaction.api;
+
+import java.util.function.Consumer;
+
+public interface TimestampLockAwareTransaction {
+    /**
+     * Similar to {@link Transaction#preCommit(Runnable)} with the option of fetching timestamps.
+     * <p>
+     * A lock will be taken on a timestamp before all timestamps provided in {@code preCommitAction}.
+     * This lock will be checked at commit time, and if expired, will fail the transaction.
+     * <p>
+     * Clients can use {@link #getLockedTimestamp(String)} to fetch the earliest locked
+     * timestamp for a given {@param timestampLockDescriptor} of an open transaction.
+     * Note these semantics are the quite similar to {@link TransactionManager#getImmutableTimestamp()}, but instead
+     * of tracking open start transaction timestamps, we track open {@code preCommitAction} timestamps.
```

Comment:
As discussed: the consequence of this API is that you EXPECT the user to just call into their stores (and by extension the Transaction in AtlasDB) so this NEEDs to happen before the code in SnapshotTransaction#commitWrites or something.

Is there a usecase that you had for this style of work outside of this workstream? Because I struggle to understand why this is better than the more generic idea of "delayed-writes" (and not allowing people to interact with their Transaction object at the point where that function is ran).

Reply:
Additional concern I'd have is: would people expect that, if they happen to use their Transaction object from multiple threads, that there's an ordering between all interactions with the transaction object happened AND THEN these runnables are ran?

So concretely, would we want to have another state in https://github.com/palantir/atlasdb/blob/develop/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java#L243 or something?